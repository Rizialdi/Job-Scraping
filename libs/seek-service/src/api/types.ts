/**
 * Base configuration for the SeekService
 */
interface ISeekServiceBaseConfig {
  baseUrl: string;
  api: {
    v4: {
      baseUrl: string;
      endpoints: {
        search: string;
        relatedSearch: string;
      };
    };
  };
}

/**
 * Job listing from the search endpoint
 */
interface ISeekServiceSearchResponseJobListing {
  advertiser: {
    id: string;
    description: string;
  };
  area?: string;
  areaId?: number;
  areaWhereValue?: string;
  automaticInclusion: boolean;
  branding?: {
    id: string;
    assets: {
      logo: {
        strategies: {
          jdpLogo: string;
          serpLogo: string;
        };
      };
    };
  };
  bulletPoints: Array<string>;
  classification: {
    id: string;
    description: string;
  };
  companyName?: string;
  companyProfileStructuredDataId: number;
  displayType: string;
  listingDateDisplay: string;
  location: string;
  locationId: number;
  locationWhereValue: string;
  id: number;
  isPremium: boolean;
  isStandOut: boolean;
  jobLocation: {
    label: string;
    countryCode: string;
    seoHierarchy: {
      contextualName: string;
    }[];
  };
  listingDate: string;
  logo: {
    id: string;
    description: unknown;
  };
  roleId: string;
  salary: string;
  solMetadata: {
    searchRequestToken: string;
    token: string;
    jobId: string;
    section: string;
    sectionRank: number;
    jobAdType: string;
    tags: {
      mordor__flights: string;
      mordor__s: string;
    };
  };
  subClassification: {
    id: string;
    description: string;
  };
  suburb?: string;
  suburbId?: number;
  suburbWhereValue?: string;
  teaser: string;
  title: string;
  tracking: string;
  workType: string;
  isPrivateAdvertiser: boolean;
  displayStyle?: {
    search?: string;
  };
}

/**
 * Response from the search endpoint
 */
interface ISeekServiceSearchResponse {
  data: ISeekServiceSearchResponseJobListing[];
  title: string;
  totalCount: number;
  paginationParameters: {
    seekSelectAllPages: boolean;
    hadPremiumListings: boolean;
  };
  info: {
    timeTaken: number;
    source: string;
    experiment: string;
  };
  userQueryId: string;
  sortMode: {
    isActive: boolean;
    name: string;
    value: string;
  }[];
  solMetadata: {
    requestToken: string;
    token: string;
    keywords: string;
    sortMode: string;
    pageSize: number;
    pageNumber: number;
    totalJobCount: number;
    tags: {
      mordor__flights: string;
      'chalice-search-api:solId': string;
    };
  };
  joraCrossLink: {
    canCrossLink: boolean;
  };
  searchParams: {
    page: string;
    keywords: string;
    solid: string;
  };
}

export {
  ISeekServiceBaseConfig,
  ISeekServiceSearchResponse,
  ISeekServiceSearchResponseJobListing,
};
