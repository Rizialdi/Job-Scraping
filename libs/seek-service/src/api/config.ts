import { ISeekServiceBaseConfig } from './types';

/**
 * Default configuration for the SeekService
 */
const defaultConfiguration: ISeekServiceBaseConfig = {
  baseUrl: 'https://www.seek.com.au',
  api: {
    v4: {
      baseUrl: '/api/chalice-search/v4',
      endpoints: {
        search: '/search?page={pageNumber}&keywords={keyword}',
        relatedSearch:
          '/related-search?zone=anz-1&siteKey=au&keywords={keyword}',
      },
    },
  },
};

export { defaultConfiguration };
