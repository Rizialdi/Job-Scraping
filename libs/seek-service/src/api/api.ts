import axios, { AxiosInstance } from 'axios';
import { defaultConfiguration } from './config';
import {
  ISeekServiceApiBaseConfig,
  ISeekServiceApiRelatedSearchResponse,
  ISeekServiceApiSearchResponse,
} from './types';

/**
 * SeekService API
 * This class is responsible for making requests to the Seek API
 */
class SeekServiceApi {
  public config: ISeekServiceApiBaseConfig;
  private axiosInstance: AxiosInstance;

  constructor() {
    this.config = defaultConfiguration;

    this.axiosInstance = axios.create({
      baseURL: this.config.baseUrl,
      timeout: 5000,
    });
  }

  /**
   * Get the current configuration for the SeekService
   * @returns The current configuration for the SeekService
   */
  public getConfiguration(): ISeekServiceApiBaseConfig {
    return this.config;
  }

  /**
   * Get job listings given a keyword and a given page number
   * @param keyword - The keyword to search for
   * @param pageNumber - The page number to search for
   * @returns The response from the API
   */
  public async search(
    keyword: string,
    pageNumber: number
  ): Promise<ISeekServiceApiSearchResponse> {
    const {
      api: {
        v4: { baseUrl, endpoints },
      },
    } = this.config;

    const url = `${baseUrl}${endpoints.search}`
      .replace('{pageNumber}', pageNumber.toString())
      .replace('{keyword}', keyword);

    const { data } = await this.axiosInstance.get(url);

    return data;
  }

  /**
   * Get related keywords given a keyword
   * @param keyword - The keyword to search for
   * @returns The response from the API
   */
  public async relatedSearch(
    keyword: string
  ): Promise<ISeekServiceApiRelatedSearchResponse> {
    const {
      api: {
        v4: { baseUrl, endpoints },
      },
    } = this.config;

    const url = `${baseUrl}${endpoints.relatedSearch}`.replace(
      '{keyword}',
      keyword
    );

    const { data } = await this.axiosInstance.get(url);

    return data;
  }
}

export { SeekServiceApi };
