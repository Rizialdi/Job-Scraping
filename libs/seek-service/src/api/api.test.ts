import * as nock from 'nock';
import { searchResult } from '../__tests__/stubs/api.stub';
import { SeekServiceApi } from './api';

describe('#SeekServiceApi', () => {
  let sut: SeekServiceApi;
  let keyword: string;
  let pageNumber: number;

  beforeEach(() => {
    sut = new SeekServiceApi();

    pageNumber = 1;
    keyword = 'fullstack';
  });

  afterEach(() => {
    nock.cleanAll();
  });

  describe('#constructor', () => {
    it('should be defined', () => {
      expect(sut).toBeDefined();
    });

    it('should get the default configuration', () => {
      expect(sut.config).toMatchSnapshot();
    });
  });

  describe('#search', () => {
    it('should call the correct endpoint', async () => {
      const scope = nock('https://www.seek.com.au')
        .get(
          `/api/chalice-search/v4/search?page=${pageNumber}&keywords=${keyword}`
        )
        .reply(200);

      await sut.search(keyword, pageNumber);

      scope.done();
    });

    it('should have the response matching the snapshot', async () => {
      const scope = nock('https://www.seek.com.au')
        .get(
          `/api/chalice-search/v4/search?page=${pageNumber}&keywords=${keyword}`
        )
        .reply(200, searchResult());

      const response = await sut.search(keyword, pageNumber);

      scope.done();
      expect(response).toMatchSnapshot();
    });
  });
});
