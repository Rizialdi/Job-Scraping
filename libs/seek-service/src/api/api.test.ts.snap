// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#SeekServiceApi #constructor should get the default configuration 1`] = `
{
  "api": {
    "v4": {
      "baseUrl": "/api/chalice-search/v4",
      "endpoints": {
        "relatedSearch": "/related-search?zone=anz-1&siteKey=au&keywords={keyword}",
        "search": "/search?page={pageNumber}&keywords={keyword}",
      },
    },
  },
  "baseUrl": "https://www.seek.com.au",
}
`;

exports[`#SeekServiceApi #search should have the response matching the snapshot 1`] = `
{
  "data": [
    {
      "advertiser": {
        "description": "Bridge Associates APAC Pty Ltd",
        "id": "28815495",
      },
      "area": "North Shore & Northern Beaches",
      "areaId": 5028,
      "areaWhereValue": "North Shore & Northern Beaches Sydney NSW",
      "automaticInclusion": false,
      "bulletPoints": [
        "Exciting New Role -",
        "Career Development",
      ],
      "classification": {
        "description": "Information & Communication Technology",
        "id": "6281",
      },
      "companyName": "Bridge Associates APAC",
      "companyProfileStructuredDataId": 221565,
      "displayStyle": {
        "search": "A",
      },
      "displayType": "promoted",
      "id": 61986871,
      "isPremium": true,
      "isPrivateAdvertiser": false,
      "isStandOut": true,
      "jobLocation": {
        "countryCode": "AU",
        "label": "North Sydney, Sydney NSW",
        "seoHierarchy": [
          {
            "contextualName": "North Sydney NSW 2060",
          },
          {
            "contextualName": "All Sydney NSW",
          },
        ],
      },
      "listingDate": "2023-02-28T12:03:01Z",
      "listingDateDisplay": "4d ago",
      "location": "Sydney",
      "locationId": 1000,
      "locationWhereValue": "All Sydney NSW",
      "logo": {
        "description": null,
        "id": "",
      },
      "roleId": "digital-lead",
      "salary": "$140,000 - $169,999",
      "solMetadata": {
        "jobAdType": "SPONSORED",
        "jobId": "61986871",
        "searchRequestToken": "840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
        "section": "MAIN",
        "sectionRank": 1,
        "tags": {
          "mordor__flights": "mordor_123",
          "mordor__s": "0",
        },
        "token": "0~840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
      },
      "subClassification": {
        "description": "Developers/Programmers",
        "id": "6287",
      },
      "suburb": "North Sydney",
      "suburbId": 24000,
      "suburbWhereValue": "North Sydney NSW 2060",
      "teaser": "Exciting new digital lead developer role working with the latest Microsoft techs!!",
      "title": "Digital Lead - Hands on Fullstack Lead  - Agile - DevOps -",
      "tracking": "ewogICJ0b2tlbiI6ICI1OTYyZjM4NS1kNDM0LTk2NGUtOTM1ZC00ZWMwZTMwZTBiNmZfMSIKfQ==",
      "workType": "Full Time",
    },
  ],
  "info": {
    "experiment": "mordor_123",
    "source": "JobSearch-sm",
    "timeTaken": 20,
  },
  "joraCrossLink": {
    "canCrossLink": false,
  },
  "paginationParameters": {
    "hadPremiumListings": true,
    "seekSelectAllPages": true,
  },
  "searchParams": {
    "keywords": "fullstack",
    "page": "1",
    "solid": "5fb082eb-8177-455f-b2a2-506c1136d09f",
  },
  "solMetadata": {
    "keywords": "fullstack",
    "pageNumber": 1,
    "pageSize": 22,
    "requestToken": "840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
    "sortMode": "RELEVANCE",
    "tags": {
      "chalice-search-api:solId": "5fb082eb-8177-455f-b2a2-506c1136d09f",
      "mordor__flights": "mordor_123",
    },
    "token": "0~840c4c61-9bf7-4a82-b8f3-cddea9a8f793",
    "totalJobCount": 670,
  },
  "sortMode": [
    {
      "isActive": true,
      "name": "Relevance",
      "value": "KeywordRelevance",
    },
    {
      "isActive": false,
      "name": "Date",
      "value": "ListedDate",
    },
  ],
  "title": "Fullstack Jobs",
  "totalCount": 670,
  "userQueryId": "API3473179588813666096",
}
`;
